lib/Optimizer/Transforms/ConstPropComplex.cpp

```diff
@@ -0,0 +1,210 @@
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(arith::ExtFOp extend,
+                                PatternRewriter &rewriter) const override {
+    auto val = extend.getOperand();
+    auto valCon = val.getDefiningOp<arith::ConstantFloatOp>();
+    if (valCon) {
+      auto v = valCon.value().convertToFloat();
+      auto fTy = dyn_cast<FloatType>(extend.getType());
+      rewriter.replaceOpWithNewOp<arith::ConstantFloatOp>(
+          extend, APFloat{static_cast<double>(v)}, fTy);
+      return success();
+    }
+    return failure();
+  }
+};
+
+// Fold complex.re ops if the argument is constant.
+class ComplexRePattern : public OpRewritePattern<complex::ReOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(complex::ReOp re,
+                                PatternRewriter &rewriter) const override {
+    auto val = re.getOperand();
+    auto valCon = val.getDefiningOp<complex::ConstantOp>();
+    if (valCon) {
+      auto attr = valCon.getValue();
+      auto real = cast<FloatAttr>(attr[0]).getValue();
+      auto fTy = dyn_cast<FloatType>(re.getType());
+      auto v = fTy.isF64() ? real.convertToDouble() : real.convertToFloat();
+      rewriter.replaceOpWithNewOp<arith::ConstantFloatOp>(re, APFloat{v}, fTy);
+      return success();
+    }
+    return failure();
+  }
+};
+
+// Fold complex.im ops if the argument is constant.
+class ComplexImPattern : public OpRewritePattern<complex::ImOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(complex::ImOp im,
+                                PatternRewriter &rewriter) const override {
+    auto val = im.getOperand();
+    auto valCon = val.getDefiningOp<complex::ConstantOp>();
+    if (valCon) {
+      auto attr = valCon.getValue();
+      auto imag = cast<FloatAttr>(attr[1]).getValue();
+      auto fTy = dyn_cast<FloatType>(im.getType());
+      auto v = fTy.isF64() ? imag.convertToDouble() : imag.convertToFloat();
+      rewriter.replaceOpWithNewOp<arith::ConstantFloatOp>(im, APFloat{v}, fTy);
+      return success();
+    }
+    return failure();
+  }
+};
+
+class ConstPropComplexPass
+    : public cudaq::opt::impl::ConstPropComplexBase<ConstPropComplexPass> {
+public:
+  using ConstPropComplexBase::ConstPropComplexBase;
+
+  void runOnOperation() override {
+    auto *ctx = &getContext();
+    auto module = getOperation();
+    for (Operation &op : *module.getBody()) {
+      auto func = dyn_cast<func::FuncOp>(op);
+      if (!func)
+        continue;
+      DominanceInfo domInfo(func);
+      std::string funcName = func.getName().str();
+      RewritePatternSet patterns(ctx);
+      patterns.insert<ComplexCreatePattern>(ctx);
+      patterns.insert<FloatExtendPattern>(ctx);
+      patterns.insert<FloatTruncatePattern>(ctx);
+      patterns.insert<ComplexRePattern>(ctx);
+      patterns.insert<ComplexImPattern>(ctx);
+      patterns.insert<CastArrayPtrPattern>(ctx);
```

Comment:
nit: These can be combined into a single line.

Reply:
done, thanks!